File: priv/templates/companies/bench-softdev/fixtures/repo/lib/comment_renderer.ex
defmodule BenchFixture.CommentRenderer do
  @moduledoc """
  Renders user comments to HTML. Bench task bugs-3 asks the
  engineer to plug the XSS hole — `render/1` currently emits raw
  HTML and the task body must be escaped.

  NOTE: intentionally vulnerable. Don't copy this module into
  production code — it's a bench fixture.
  """

  @doc """
  Render a comment map (`%{body: binary, author: binary}`) to
  HTML. Current implementation passes `body` through raw.
  """
  def render(%{body: body, author: author}) do
    "<article><header>@#{author}</header><section>#{body}</section></article>"
  end

  @doc """
  Wrap a URL into an `<a href>` link. Callers pre-compute this;
  the sanitizer added by bugs-3 must preserve these links.
  """
  def wrap_link(url, text) when is_binary(url) and is_binary(text) do
    {:safe, ~s(<a href="#{url}">#{text}</a>)}
  end
end

File: priv/templates/companies/bench-softdev/fixtures/repo-py/src/comment_renderer.py
"""Render user comments to HTML.

bench task bugs-py-3 asks the engineer to plug the XSS hole —
``render()`` currently emits raw HTML and the task body must be
escaped.

NOTE: intentionally vulnerable. Do not copy this module into
production code — it is a bench fixture.
"""

from __future__ import annotations

from typing import Mapping


def render(comment: Mapping[str, str]) -> str:
    body = comment["body"]
    author = comment["author"]
    return f"<article><header>@{author}</header><section>{body}</section></article>"


def wrap_link(url: str, text: str) -> tuple[str, str]:
    """Wrap a URL into an ``<a href>`` link.

    Callers pre-compute this; the sanitizer added by bugs-py-3
    must preserve these links.
    """
    return ("safe", f'<a href="{url}">{text}</a>')
